packages/xstate-test/src/TestModel.ts

```diff
@@ -0,0 +1,491 @@
+    statePredicate: StatePredicate<TState>
+  ): StatePlan<TState, TEvent> {
+    const path = getPathFromEvents(this.behavior, events);
+
+    if (!statePredicate(path.state)) {
+      throw new Error(
+        `The last state ${JSON.stringify(
+          (path.state as any).value
+        )} does not match the target}`
+      );
+    }
+
+    const plan: StatePlan<TState, TEvent> = {
+      state: path.state,
+      paths: [path]
+    };
+
+    return plan;
+  }
+
+  public getAllStates(): TState[] {
+    const adj = performDepthFirstTraversal(this.behavior, this.options);
+    return Object.values(adj).map((x) => x.state);
+  }
+
+  public async testPlan(
+    plan: StatePlan<TState, TEvent>,
+    options?: Partial<TestModelOptions<TState, TEvent>>
+  ) {
+    for (const path of plan.paths) {
+      await this.testPath(path, options);
+    }
+  }
+
+  public testPlanSync(
+    plan: StatePlan<TState, TEvent>,
+    options?: Partial<TestModelOptions<TState, TEvent>>
+  ) {
+    for (const path of plan.paths) {
+      this.testPathSync(path, options);
+    }
+  }
+
+  public testPathSync(
+    path: StatePath<TState, TEvent>,
+    options?: Partial<TestModelOptions<TState, TEvent>>
+  ) {
+    const testPathResult: TestPathResult = {
+      steps: [],
+      state: {
+        error: null
+      }
+    };
+
+    try {
+      for (const step of path.steps) {
+        const testStepResult: TestStepResult = {
+          step,
+          state: { error: null },
+          event: { error: null }
+        };
+
+        testPathResult.steps.push(testStepResult);
+
+        try {
+          this.testStateSync(step.state, options);
+        } catch (err) {
+          testStepResult.state.error = err;
+
+          throw err;
+        }
+
+        try {
+          this.testTransitionSync(step);
+        } catch (err) {
+          testStepResult.event.error = err;
+
+          throw err;
+        }
+      }
+
+      try {
+        this.testStateSync(path.state, options);
+      } catch (err) {
+        testPathResult.state.error = err.message;
+        throw err;
+      }
+    } catch (err) {
+      // TODO: make option
+      err.message += formatPathTestResult(path, testPathResult, this.options);
+      throw err;
+    }
+  }
+
+  public async testPath(
+    path: StatePath<TState, TEvent>,
+    options?: Partial<TestModelOptions<TState, TEvent>>
+  ) {
+    const testPathResult: TestPathResult = {
+      steps: [],
+      state: {
+        error: null
+      }
+    };
+
+    try {
+      for (const step of path.steps) {
+        const testStepResult: TestStepResult = {
+          step,
+          state: { error: null },
+          event: { error: null }
+        };
+
+        testPathResult.steps.push(testStepResult);
+
+        try {
+          await this.testState(step.state, options);
+        } catch (err) {
+          testStepResult.state.error = err;
+
+          throw err;
+        }
+
+        try {
+          await this.testTransition(step);
```

Comment:
re duplication between `testPlan` and `testPlanSync` etc... the main (and only?) difference is that `testPlan` automatically awaits some results, right? so... what if we'd just check if the result is a promise and conditionally await it? This way we could just have a single implementation.

The outermost call would still have to be an async function though to use `await` conveniently in the implementation **if** the return value turns out to be a promise. OTOH we could technically ditch `await` sugar and deal with bare promises here in an attempt to have just a single implementation.

The inner calls wouldn't be able to use `async` sugar either because even if they don't return anything their void return value would still be promisified. The whole async call stack is just "colored" this way and we can't escape it within the bounds of the sugar syntax.

We could ditch the sugar altogether and it's totally doable **but** it complicates the control flow with loops, try/catches etc. Ultimately - it wouldn't quite be worth it.

There is a neat language construct that would actually allow us to share the implementation, has a nice sugar for the control flow, and that allows a function~ to return any value, that doesn't mandate returning a promise... and that construct is called a generator function 😅 

Generators might not always be super ergonomic for public consumption but it's usually fairly easy to write a wrapper function that calls into the generator and return a plain value for the caller. This is actually how some methods in the Babel are implemented nowadays (for example you have `transformFile` and `transformFileSync` APIs that share the very same, generator-based, implementation, but that's completely opaque to the user). They have even made a package that makes implementing this pattern easier: https://github.com/loganfsmyth/gensync

It's probably not worth at all to refactor this at all right now, I'm just laying out the possibilities here. 

Generators have a nice trait though - they are lazy by default, you could have a plan generator that would generate a *single* path, execute it then and only generate another plan after you are done with the first one. We could even implement test parallelization at this level~ without paying the price to generate *all* the paths upfront which gives us somewhat of a cold start before any test can even be executed.

(sorry for the chaos, was just thinking out loud, in writing)

Reply:
cc. @mattpocock What do you think?

Reply:
Note that the lazy generation probably wouldnt work for plans that require knowing all of the generated paths for weight-based filtering and stuff - but I wonder if that's not the limitation of the depth-first traversal that we use. It feels like with breadth-first traversal it could be possible to do this

Reply:
I agree with Andarist, I don't mind this duplicated code since it's likely easier to read than the alternative proposed above.